extensions/csrc/cuda/attention/attention_utils.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Adapted from
+ * https://github.com/NVIDIA/FasterTransformer/blob/release/v5.3_tag/src/fastertransformer/kernels/decoder_masked_multihead_attention/decoder_masked_multihead_attention_template.hpp
+ * Copyright (c) 2024, The Colossal-AI team.
+ * Copyright (c) 2023, The vLLM team.
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+#include <float.h>
+
+#include "../funcs/binary_functor.h"
+#include "../funcs/cast_functor.h"
+#include "../funcs/ternary_functor.h"
+#include "../funcs/unary_functor.h"
+#include "../utils/vec_type_traits.h"
+
+namespace colossalAI {
+namespace cuda {
+namespace attention {
+
+using colossalAI::cuda::funcs::BinaryOpFunctor;
+using colossalAI::cuda::funcs::BinaryOpType;
+using colossalAI::cuda::funcs::TernaryOpFunctor;
+using colossalAI::cuda::funcs::TernaryOpType;
+using colossalAI::cuda::funcs::UnaryOpFunctor;
+using colossalAI::cuda::funcs::UnaryOpType;
+using colossalAI::cuda::utils::FloatVecTypeTrait;
+
+#define WARP_SIZE 32
+#define VEC_SIZE_8 8
+
+#define SHFL_XOR_SYNC(var, lane_mask) \
+  __shfl_xor_sync(uint32_t(-1), var, lane_mask)
+#define SHFL_SYNC(var, src_lane) __shfl_sync(uint32_t(-1), var, src_lane)
+
+// Q*K^T operation.
+template <int NUM_THREADS_PER_TOKEN, typename Vec, int N>
```

Comment:
VecT

Reply:
So is there exists any problem?

Reply:
Vec -> VecT I means, maybe it is more clear?